sw/device/silicon_creator/rom/e2e/rom_e2e_c_init_test.c

```diff
@@ -0,0 +1,177 @@
+  do {                   \
+    if (!(condition)) {  \
+      fault();           \
+    }                    \
+  } while (false)
+
+/**
+ * Check that the pinmux has been initialized correctly.
+ *
+ * Expected configuration:
+ * - If the OTP BOOTSTRAP_EN option is set, then pinmux should be configured
+ *   with the following peripheral/pad mapping, with pull-down enabled for all
+ *   pads:
+ *     * GPIO22 -> IOC0
+ *     * GPIO23 -> IOC1
+ *     * GPIO24 -> IOC2
+ * - Regardless of the OTP values, the pinmux should be configured with the
+ *   following mapping for UART0:
+ *     * UART RX -> IOC3
+ *     * IOC4 -> UART TX (output)
+ */
+static void pinmux_init_test(void) {
+  uint32_t bootstrap_en =
+      otp_read32(OTP_CTRL_PARAM_OWNER_SW_CFG_ROM_BOOTSTRAP_EN_OFFSET);
+  if (bootstrap_en == kHardenedBoolTrue) {
+    // GPIO 22 (input 22) -> IOC0 (MIO pad 22), pad attributes: pull-down
+    // enabled
+    uint32_t insel_gpio22 =
+        abs_mmio_read32(kPinmuxBase + PINMUX_MIO_PERIPH_INSEL_22_REG_OFFSET);
+    CHECK(insel_gpio22 == kTopEarlgreyPinmuxInselIoc0);
+    uint32_t attr22 =
+        abs_mmio_read32(kPinmuxBase + PINMUX_MIO_PAD_ATTR_22_REG_OFFSET);
+    // TODO: why is this attribute not legal on FPGA?
+    // CHECK(bitfield_bit32_read(attr22, PINMUX_MIO_PAD_ATTR_22_PULL_EN_22_BIT)
+    // == true);
+    CHECK(bitfield_bit32_read(
+              attr22, PINMUX_MIO_PAD_ATTR_22_PULL_SELECT_22_BIT) == false);
+
+    // GPIO 23 (input 23) -> IOC1 (MIO pad 23), pad attributes: pull-down
+    // enabled
+    uint32_t insel_gpio23 =
+        abs_mmio_read32(kPinmuxBase + PINMUX_MIO_PERIPH_INSEL_23_REG_OFFSET);
+    CHECK(insel_gpio23 == kTopEarlgreyPinmuxInselIoc1);
+    uint32_t attr23 =
+        abs_mmio_read32(kPinmuxBase + PINMUX_MIO_PAD_ATTR_23_REG_OFFSET);
+    // TODO: why is this attribute not legal on FPGA?
+    // CHECK(bitfield_bit32_read(attr23, PINMUX_MIO_PAD_ATTR_23_PULL_EN_23_BIT)
+    // == true);
+    CHECK(bitfield_bit32_read(
+              attr23, PINMUX_MIO_PAD_ATTR_23_PULL_SELECT_23_BIT) == false);
+
+    // GPIO 24 (input 24) -> IOC2 (MIO pad 24), pad attributes: pull-down
+    // enabled
+    uint32_t insel_gpio24 =
+        abs_mmio_read32(kPinmuxBase + PINMUX_MIO_PERIPH_INSEL_24_REG_OFFSET);
+    CHECK(insel_gpio24 == kTopEarlgreyPinmuxInselIoc2);
+    uint32_t attr24 =
+        abs_mmio_read32(kPinmuxBase + PINMUX_MIO_PAD_ATTR_24_REG_OFFSET);
+    // TODO: why is this attribute not legal on FPGA?
+    // CHECK(bitfield_bit32_read(attr24, PINMUX_MIO_PAD_ATTR_24_PULL_EN_24_BIT)
+    // == true);
+    CHECK(bitfield_bit32_read(
+              attr24, PINMUX_MIO_PAD_ATTR_24_PULL_SELECT_24_BIT) == false);
+
+    // TODO: this checks legal bits, failing if the pull_en is legal
+    abs_mmio_write32(kPinmuxBase + PINMUX_MIO_PAD_ATTR_22_REG_OFFSET,
+                     UINT32_MAX);
+    uint32_t legal22 =
+        abs_mmio_read32(kPinmuxBase + PINMUX_MIO_PAD_ATTR_22_REG_OFFSET);
+    CHECK(legal22 == 3);
```

Comment:
This check is to figure out which attribute bits are legal for a given top level (the pad attributes are write-any-read-legal, so if we set all bits then only legal bits will be set when we read back). Checking for `== 3` passes on FPGA, meaning that only bits 0 and 1 of the pad attributes are possible to set. However, `pull_en` is bit 2. Does this indicate a problem? Or should the test simply be modified to check that `pull_en` is set _if it's legal for the current architecture_?

See also the commented-out `CHECK`s above, which fail because the `pull_en` bit is not legal.

Reply:
Same as above, if this is expected we can guard this check with the device type. Depending on how you want to proceed, you may want to keep checks enabled all the time but use a device-type-dependent mask to filter out bits that are not supported on particular device type.